hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/merge/SparkRecordMergingUtils.java

```diff
@@ -0,0 +1,192 @@
+      StructType structType = HoodieInternalRowUtils.getCachedSchema(schema);
+      Map<Integer, StructField> schemaFieldIdMapping = new HashMap<>();
+      int fieldId = 0;
+
+      for (StructField field : structType.fields()) {
+        schemaFieldIdMapping.put(fieldId, field);
+        fieldId++;
+      }
+
+      return schemaFieldIdMapping;
+    });
+  }
+
+  /**
+   * @param avroSchema Avro schema.
+   * @return The field name to ID mapping.
+   */
+  public static Map<String, Integer> getCachedFieldNameToIdMapping(Schema avroSchema) {
+    return FIELD_NAME_TO_ID_MAPPING_CACHE.computeIfAbsent(avroSchema, schema -> {
+      StructType structType = HoodieInternalRowUtils.getCachedSchema(schema);
+      Map<String, Integer> schemaFieldIdMapping = new HashMap<>();
+      int fieldId = 0;
+
+      for (StructField field : structType.fields()) {
+        schemaFieldIdMapping.put(field.name(), fieldId);
+        fieldId++;
+      }
+
+      return schemaFieldIdMapping;
+    });
+  }
+
+  /**
+   * Merges the two schemas so the merged schema contains all the fields from the two schemas,
+   * with the same ordering of fields, and the fields of the old schema comes first.
+   *
+   * @param oldSchema Old schema.
+   * @param newSchema New schema.
+   * @return The ID to {@link StructField} instance mapping of the merged schema, and the
+   * {@link StructType} and Avro schema of the merged schema.
+   */
+  public static Pair<Map<Integer, StructField>, Pair<StructType, Schema>> getCachedMergedSchema(Schema oldSchema,
+                                                                                                Schema newSchema) {
+    return MERGED_SCHEMA_CACHE.computeIfAbsent(Pair.of(oldSchema, newSchema), schemaPair -> {
+      Schema schema1 = schemaPair.getLeft();
+      Schema schema2 = schemaPair.getRight();
+      Map<Integer, StructField> idToFieldMapping1 = getCachedFieldIdToFieldMapping(schema1);
+      Map<Integer, StructField> idToFieldMapping2 = getCachedFieldIdToFieldMapping(schema2);
+      Map<String, Integer> nameToIdMapping1 = getCachedFieldNameToIdMapping(schema1);
+      Map<String, Integer> nameToIdMapping2 = getCachedFieldNameToIdMapping(schema2);
+      List<Integer> newFieldIdList = new ArrayList<>();
+      for (String name : nameToIdMapping2.keySet()) {
+        if (!nameToIdMapping1.containsKey(name)) {
+          newFieldIdList.add(nameToIdMapping2.get(name));
+        }
+      }
+
+      if (newFieldIdList.isEmpty()) {
+        return Pair.of(idToFieldMapping1, Pair.of(HoodieInternalRowUtils.getCachedSchema(oldSchema), schema1));
+      } else {
+        Map<Integer, StructField> mergedMapping = new HashMap<>(idToFieldMapping1);
+        int newFieldId = mergedMapping.size();
+        newFieldIdList.sort(Comparator.naturalOrder());
+        for (Integer fieldId : newFieldIdList) {
+          mergedMapping.put(newFieldId, idToFieldMapping2.get(fieldId));
+          newFieldId++;
```

Comment:
The `newFieldId` is just a logical id for the differ new fields, we do not allow dropping fields right?

Reply:
Correct.